gobblin-hive-registration/src/main/java/gobblin/hive/HiveRegister.java

```diff
@@ -0,0 +1,302 @@
+          addPartition(table, partition.get());
+        }
+
+        if (spec instanceof HiveSpecWithPostActivities) {
+          for (Activity<HiveMetaStoreClient> activity : ((HiveSpecWithPostActivities) spec).getActivities()) {
+            activity.execute(HiveRegister.this.client);
+          }
+        }
+
+        return null;
+      }
+    });
+
+    this.futures.add(future);
+    return future;
+  }
+
+  private boolean evaluatePredicates(HiveSpecWithPredicates spec) {
+    for (Predicate<HiveMetaStoreClient> pred : spec.getPredicates()) {
+      if (!pred.apply(this.client)) {
+        return false;
+      }
+    }
+    return true;
+  }
+
+  private void createDbIfNotExists(String dbName) {
+    Database db = new Database();
+    db.setName(dbName);
+
+    Preconditions.checkState(this.hiveDbRootDir.isPresent(),
+        "Missing required property " + HiveRegProps.HIVE_DB_ROOT_DIR);
+    db.setLocationUri(this.hiveDbRootDir.get() + Path.SEPARATOR + dbName + HIVE_DB_EXTENSION);
+
+    Lock lock = locks.get(dbName);
+    lock.lock();
+
+    try {
+      client.createDatabase(db);
+    } catch (AlreadyExistsException e) {
+      log.warn("Hive database " + dbName + " already exists");
+    } catch (TException e) {
+      log.error("Unable to create Hive database " + dbName);
+      throw Throwables.propagate(e);
+    } finally {
+      lock.unlock();
+    }
+  }
+
+  private Table getOrCreateTable(String dbName, String tableName, Optional<HivePartition> partition,
+      StorageDescriptor sd) {
+    Lock lock = locks.get(dbName + tableName);
+    lock.lock();
+
+    try {
+      return this.client.getTable(dbName, tableName);
+    } catch (NoSuchObjectException e) {
+      return createTable(dbName, tableName, partition, sd);
+    } catch (TException e) {
+      log.error(String.format("Unable to check existence of Hive table %s in db %s", tableName, dbName));
+      throw Throwables.propagate(e);
+    } finally {
+      lock.unlock();
+    }
+  }
+
+  private Table createTable(String dbName, String tableName, Optional<HivePartition> partition, StorageDescriptor sd) {
+    Table table = new Table();
+    table.setParameters(Maps.<String, String> newHashMap());
+    table.getParameters().put("EXTERNAL", Boolean.TRUE.toString());
+    table.setDbName(dbName);
+    table.setTableName(tableName);
+    if (partition.isPresent()) {
+      table.setPartitionKeys(partition.get().getKeys());
+    }
+
+    Preconditions.checkState(this.hiveOwner.isPresent(), "Missing required property " + HiveRegProps.HIVE_OWNER);
+    table.setOwner(this.hiveOwner.get());
+    table.setTableType("EXTERNAL_TABLE");
+    table.setSd(sd);
+    try {
+      this.client.createTable(table);
+    } catch (TException e) {
+      log.error(String.format("Unable to create table %s from db %s", tableName, dbName));
+      throw Throwables.propagate(e);
+    }
+    return table;
+  }
+
+  public void deregister(String dbName, String tableName) {
```

Comment:
Method name sounds a little value, deregistering what exactly? Recommend changing to `deregisterTable` or `dropTable`. Similar comment for the overloaded methods.